### PR TITLE
`parse_shapelike` allows 0

### PR DESCRIPTION
**Pull request by d-v-b**
- `parse_shapelike` allows shape values to be 0
- `parse_shapelike` allows the empty tuple `()`
- `parse_shapelike` rejects negative integers correctly (in `main` this is not working)
- expands tests accordingly

response to #1976 and #1977, although I'm not sure it addresses those concerns entirely, since I haven't changed array indexing tests at all

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

### CODE REVIEW

File: src/zarr/common.py [hunk 1]
```diff
@@ -137,17 +137,21 @@ def parse_named_configuration(
...
+        if data < 0:
+            raise ValueError(f"Expected a non-negative integer. Got {data} instead")
         return (data,)
-    if not isinstance(data, Iterable):
-        raise TypeError(f"Expected an iterable. Got {data} instead.")
-    data_tuple = tuple(data)
-    if len(data_tuple) == 0:
-        raise ValueError("Expected at least one element. Got 0.")
+    try:
+        data_tuple = tuple(data)
```
Comment (jhamman) [inline src/zarr/common.py:144]: we should also be checking that all values are integers here. NumPy raises an error if you pass a float as a shape

```
In [4]: np.zeros((4., ))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[4], line 1
----> 1 np.zeros((4., ))

TypeError: 'float' object cannot be interpreted as an integer
```

File: tests/v3/test_common.py [hunk 2]
```diff
@@ -48,47 +48,71 @@ def test_parse_name_invalid(data: tuple[Any, Any]):
...
+def test_parse_shapelike_invalid_iterable_types(data: Any) -> None:
+    """
+    Test that we get the expected error message when passing in an iterable containing
+    non-integer elements
+    """
+    with pytest.raises(TypeError, match="Expected an iterable of integers"):
+        parse_shapelike(data)
+
+
+@pytest.mark.parametrize("data", [(1, 2, 3, -1), (-10,)])
```
Comment (jhamman) [inline tests/v3/test_common.py:95]:
```suggestion
@pytest.mark.parametrize("data", [(1, 2, 3, -1), (-10,), (4.0, 2)])
```

**Review by jhamman — COMMENTED** [review]
looks good, left you a few small suggestions.

File: src/zarr/common.py (hunk 1 above)
Comment (d-v-b) [reply src/zarr/common.py:144]: This is addressed by [32ba006](https://github.com/zarr-developers/zarr-python/pull/1979/commits/32ba006334eeb7a746978d65af18d222f0ff9d21) and [e866f24](https://github.com/zarr-developers/zarr-python/pull/1979/commits/e866f247877c5cea462709e3d8fa1f8741ef59e0)

**Pull request merged**